doc/neps/nep-0043-extensible-ufuncs.rst

```diff
@@ -0,0 +1,631 @@
+
+Alternatives and details to the ``resolve_descriptors`` function are described
+below.
+
+The current signature must be extended to allow the return value, as well
+as error reporting.  The choice of passing in a pointer to a struct means
+minimal adjustments to current functions which do not require it (they only
+need a ``0`` return value).  It may also simplify the addition of future
+parameters if necessary.
+
+The main alternative would be extending the signature either by only a
+return value giving a much higher burden to implement a user setup.
+
+
+Reusing existing Loops/Implementations
+""""""""""""""""""""""""""""""""""""""
+
+*TODO: Do we need to make both easy? Or maybe even neither? I personally prefer
+the design of having a class factory here (point 1).  I realize that many
+users seem to find the second version.*
+
+For many DTypes adding additional C-level (or python level) loops will be
+sufficient and require no more than a single strided loop implementation.
+Everything else can be provided by NumPy.  If the loop works with
+parametric DTypes, the ``resolve_descriptors`` function *must* additionally
+be provided.
+
+However, in some use-cases it is desired to call back to an existing loop.
+In Python, this can be achieved by simply calling into the original ufunc
+(when parametric types are involved potentially twice, due to calling one
+more time from ``resolve_descriptors``).
+
+For better performance in C, and for large arrays, it is desirable to reuse
+an existing implementation fully, so that the inner-loop does not need
+to be wrapped.
+The preferred solution to this is allow one additional method of setting
+up a UFunc loop conveniently without the need of directly exposing the
+the inner structure immediately.
+
+There are two potential settings here, shown with two different approaches
+for implementing a ``Unit`` DType.
+
+1. An abstract Unit DType with subclasses for each numeric dtype (these
+   could be dynamically created).  Such a ``Unit[Float64]("m")`` instance
+   has a well defined signature with respect to type promotion.
+2. A concrete Unit DType which parametrizes the numeric dtype making it
+   ``Unit(float64, "m")``.
+
+The difference between the two, is that in the first case a ``UFuncImpl``
+can be created which *wraps* a single existing ``UFuncImpl`` and after that
+dispatching can be cached.  The creation of this ``UFuncImpl`` can further
+occur dynamically in the *promotion* step.
+To fall back to the existing loop, we allow creation of a wrapped loop
+from an existing (bound) ``UFuncImpl``.
+
+This wrapped loop will have two additional methods:
+
+* ``view_inputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  user input descriptors with descriptors matching the wrapped loop.
+  It must be possible to *view* the inputs as the output.
+* ``wrap_outputs(Tuple[DType]: input_descr) -> Tuple[DType]`` replacing the
+  resolved descriptors with with the desired actual loop descriptors.
```

Comment:
I was wondering how this handles a case like np.add(Unit[Float64]("m"), Unit[Float64["km"]) ?

Reply:
Thanks, good point, I should explain that/give an example. For Units this uses `.view(np.float64)` in the inner loop effectively. In the outer loop, the descriptors will be set to `(Unit[Float64]("m"), Unit[Float64["m"], Unit[Float64]("m"))`, so the casting machinery has to convert the `"km"` to `"m"`.